okhttp/src/main/java/okhttp3/internal/connection/ExchangeFinder.kt

```diff
@@ -113,13 +113,23 @@ class ExchangeFinder(
           connectionRetryEnabled = connectionRetryEnabled
       )
 
-      // Confirm that the connection is good. If it isn't, take it out of the pool and start again.
-      if (!candidate.isHealthy(doExtensiveHealthChecks)) {
-        candidate.noNewExchanges()
-        continue
+      // Confirm that the connection is good.
+      if (candidate.isHealthy(doExtensiveHealthChecks)) {
+        return candidate
       }
 
-      return candidate
+      // If it isn't, take it out of the pool.
+      candidate.noNewExchanges()
+
+      // Make sure we have some routes left to try. One example where we may exhaust all the routes
+      // would happen if we made a new connection and it immediately is detected as unhealthy.
+      synchronized(connectionPool) {
+        val routesLeft = routeSelection?.hasNext() ?: true
+        val routesSelectionLeft = routeSelector?.hasNext() ?: true
+        if (nextRouteToTry == null && !routesLeft && !routesSelectionLeft) {
+          throw IOException("exhausted all routes")
```

Comment:
We can probably be more expressive in this error, it's an interesting case when routes are immediately unhealthy.